Refactor code related to bitvector type min/max values

antlechner opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->
This PR includes no functional changes, only refactoring.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- n/a My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

antlechner committed:
De-duplicate logic for bitvector max/min values

We should only store these arithmetic calculations in one place in the
codebase.

antlechner committed:
Move methods from (un)signedbv typet to parent class

These methods are now all identical and can be inherited from
integer_bitvector_typet.

antlechner committed:
Make use of bitvector_typet::largest

The method can be used in several places that previously duplicated its
logic.

martin-cs reviewed (approved):
Seems reasonable and code reduction patches are always a good thing!

tautschnig commented on src/util/std_types.h:
```diff
@@ -1071,6 +1071,12 @@ class bitvector_typet : public typet
+  mp_integer largest() const;
+  constant_exprt smallest_expr() const;
+  constant_exprt zero_expr() const;
+  constant_exprt largest_expr() const;
```
This is brilliant refactoring, thank you! Yet can you please hide these in `pointer_typet`? A `pointer_typet` happens to be a `bitvector_typet`, but really they must not be used (not even `zero_expr`) on a `pointer_typet`.

antlechner replied:
Ah, I overlooked that there are actually quite a few classes that inherit from `bitvector_typet`. Besides the three that were already mentioned, there are `bv_typet`, `fixedbv_typet`, `floatbv_typet`, `c_bit_field_typet` and `c_bool_typet`. However these functions will hit an `UNREACHABLE` if the type id is anything other than `ID_unsignedbv` or `ID_signedbv`. Would it make sense to introduce a new class that inherits from `bitvector_typet` and that both `signedbv_typet` and `unsignedbv_typet` inherit from? Any suggestions for what such a class should be called?

kroening replied:
smallest()/largest() only make sense for the integer ones -- so that could be integer_bitvector_typet. smallest_expr, zero_expr, largest_expr() also work on fixedbv and floatbv; these could be a numeric_bitvector_typet.

antlechner replied:
That makes sense, thank you! I only created the `integer_bitvector_typet` class as all five methods were previously only defined for the signedbv/unsignedbv types, and I added some documentation and a comment mentioning that the methods returning an exprt would make sense for fixedbv/floatbv as well in case we need them there in the future.

tautschnig commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -458,7 +446,7 @@ void initialize_nondet_string_fields(
 
   // assume (length_expr <= max_input_length)
-  if(max_nondet_string_length <= max_value(length_expr.type()))
+  if(max_nondet_string_length <= bv_spect(length_expr.type()).max_value())
```
How about `to_bitvector_type(length_expr.type()).largest()`?

antlechner replied:
Good point, why does the class `bv_spect` even (publicly) exist if this functionality is already made available by the typet?
I changed it to your suggestion for now. Depending on what you think about my previous question I might change `to_bitvector_type` to something more specific.
I also found another place where `bv_spect` was used to get the max value, and replaced this with `bitvector_type::largest` as well.

antlechner replied:
Following @kroening 's comment above, this is now `to_integer_bitvector_type(...)` (and similarly for your other comment).

tautschnig commented on src/ansi-c/ansi_c_entry_point.cpp:
```diff
@@ -377,18 +378,7 @@ bool generate_ansi_c_start_function(
-        }
-        else
-          UNREACHABLE;
+        const mp_integer max = bv_spect(envp_size_symbol.type).max_value();
```
As above `to_bitvector_type(envp_size_symbol.type).largest()`?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 23c17f1f6a97547f5cf840661bf77e59d8b2e30b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103691044

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c8f84588a66faca5c07ecf36423b15264e551e59).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103716955

antlechner committed:
Create common parent of (un)signedbv_typet classes

Since unsignedbv_typet and signedbv_typet now have some common
functionality, it makes sense to give them a common parent to store
these methods in.

antlechner committed:
Add documentation to smallest/largest/zero methods

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f313fdf5ea16757a64bbd3fa3761b17aeb0f596b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103736558
